[clang][Sema] Warn consecutive builtin comparisons in an expression

JOE1994 opened the pull request:
Add warning under `-Wparentheses` for consistency with `gcc 14.1`.

Closes #20456

JOE1994 committed:
[clang][Sema] Warn consecutive builtin comparisons in an expression

Made the following decisions for consistency with `gcc 14.1`:
* Add warning under -Wparentheses
* Set the warning to DefaultIgnore, although -Wparentheses is enabled by default
  * This warning is only issued when -Wparentheses is explicitly enabled
    (via -Wparentheses or -Wall)

Closes #20456

AaronBallman commented:
> Set the warning to DefaultIgnore, although -Wparentheses is enabled by default 

Do we know why GCC elected to go that route? This seems like something that should be enabled by default because the false positive rate should be quite low.

JOE1994 commented:
> Do we know why GCC elected to go that route? This seems like something that should be enabled by default because the false positive rate should be quite low.

It seems like `gcc` doesn't enable `-Wparentheses` by default, while `clang` enables it by default.

I will add a follow-up commit to remove `DefaultIgnore` and update affected Clang tests :)

AaronBallman commented on clang/test/Sema/parentheses.cpp:
```diff
@@ -215,3 +215,10 @@ namespace PR20735 {
+  (void)(x < y < z);  // expected-warning {{comparisons like 'X<=Y<=Z' don't have their mathematical meaning}}
+  (void)(x < y > z);  // expected-warning {{comparisons like 'X<=Y<=Z' don't have their mathematical meaning}}
+  (void)(x < y <= z); // expected-warning {{comparisons like 'X<=Y<=Z' don't have their mathematical meaning}}
+  (void)(x <= y > z); // expected-warning {{comparisons like 'X<=Y<=Z' don't have their mathematical meaning}}
```
I'd like an additional test along the lines of:
```
(void)((x < y) >= z);
```
and what about something along these lines?
```
struct S {
  bool operator<(const S &) const;
} s, t;

(void)(s < t >= z);
```
perhaps that should diagnose while this should not?
```
struct S {
  S operator<(const S &) const;
} a, b;

struct T {
  friend bool operator>=(const S &, const T &);
} c;

bool val = (a < b >= c);
```

JOE1994 replied:
(I'll refer to each of the 3 cases mentioned in the above comment as case1, case2, and case 3)

I pushed a commit that adds 2 `no-warning` test lines for case1.
It's unlikely for one to write `((x < y) >= z)` and believe it to have mathematical meaning of `(x < y) AND (y >= z)`.

I limited the scope of this PR to diagnosing consecutive "built-in" comparisons (as per the PR title).
Thus, the current implementation does not diagnose case 2.

I don't have the capacity to further work on making clang diagnose case 2 but not case 3.
I added a TODO note at the bottom of the test lines for now.

FYI, gcc doesn't diagnose any of case1, case2, or case 3.

AaronBallman replied:
> I don't have the capacity to further make clang diagnose case 2 but not case 3.
> I added a TODO note at the bottom of the test lines for now.

SGTM!

JOE1994 committed:
Update clang/lib/Sema/SemaExpr.cpp

Co-authored-by: Aaron Ballman <aaron@aaronballman.com>

JOE1994 committed:
Remove DefaultIgnore + clang-format

JOE1994 committed:
Add no-warning test lines + TODO note

Don't issue warning for expressions like `(X < Y) < Z`.  It is unlikely for one
to write `(X < Y) < Z` and confuse it to have mathematical meaning of `(X<Y) &&
(Y<Z)`.

AaronBallman commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -487,6 +487,9 @@ Improvements to Clang's diagnostics
        }
      };
 
+- Clang emits a ``-Wparentheses`` warning for expressions with consecutive comparisons like ``x < y < z``.
+  It was made a ``-Wparentheses`` warning to be consistent with gcc.
```
```suggestion
- Clang emits a ``-Wparentheses`` warning for expressions with consecutive comparisons like ``x < y < z``.
  Fixes #GH20456.
```

AaronBallman reviewed (approved):
LGTM aside from a small nit with the release notes.

JOE1994 committed:
Update clang/docs/ReleaseNotes.rst

Co-authored-by: Aaron Ballman <aaron@aaronballman.com>

erichkeane commented on clang/test/Sema/bool-compare.c:
```diff
@@ -85,7 +85,7 @@ void f(int x, int y, int z) {
 
   if ((a<y) == z) {} // no warning
-  if (a>y<z)      {} // no warning
+  if (a>y<z)      {} // expected-warning {{comparisons like 'X<=Y<=Z' don't have their mathematical meaning}}
```
I realize this is GCC's diagnostic, but I think it isn't particularly good of a diag.  I would prefer we:

1- At least match operators.  I REALLY wish we could print the actual expression (unless the expression complicated, at which point we print `<expr>` or something).  So something like: `comparison 'a>y<z' does not have its mathematical meaning`.

2- I REALLY wish we had a note after it with the range of the LHS here that says something like `Left comparison operator evaluates to `bool` before 'less-than' comparison with 'z'`.

Endilll replied:
Just throwing a wording idea: `chaining of comparison operators`.

shafik commented on clang/test/SemaTemplate/typo-template-name.cpp:
```diff
@@ -36,7 +36,7 @@ namespace InExpr {
     foo<foo; // expected-warning {{self-comparison}}
-    foo<int()>(0);
+    foo<int()>(0); // expected-warning {{comparisons like 'X<=Y<=Z' don't have their mathematical meaning}}
     foo<int(), true>(false);
```
It is a shame we don't catch this one but neither does gcc
